Concurrent `zkg create` operations can race on update of default package source

While working on zeek/package-template#4 I noticed that multiple concurrent `zkg create` operations can race on update of the default package source. I did not check which other operations can trigger such an update.

I would have expected that an operation which does not obviously mutate zkg state does properly protect that state, e.g., with file locks.

Example:
```make
# Save as `Makefile` in some directory.

# Run as many concurrent jobs as there are create operations.
MAKEFLAGS=-j6
all: 1.pkg 2.pkg 3.pkg 4.pkg 5.pkg 6.pkg

%.pkg:
	zkg create --packagedir=$@ --user-var name=Name --force
```

```console
$ make -B
zkg create --packagedir=4.pkg --user-var name=Name --force
zkg create --packagedir=5.pkg --user-var name=Name --force
zkg create --packagedir=6.pkg --user-var name=Name --force
Traceback (most recent call last):
  File "/Users/bbannier/.local/pipx/venvs/zkg/lib/python3.10/site-packages/zeekpkg/_util.py", line 60, in make_symlink
    os.symlink(target_path, link_path)
FileExistsError: [Errno 17] File exists: 'packages.zeek' -> '/Users/bbannier/.zkg/script_dir/packages/__load__.zeek'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/bbannier/.local/bin/zkg", line 2770, in <module>
    main()
  File "/Users/bbannier/.local/bin/zkg", line 2765, in main
    manager = create_manager(args, config)
  File "/Users/bbannier/.local/bin/zkg", line 387, in create_manager
    manager = zeekpkg.Manager(state_dir=state_dir, script_dir=script_dir,
  File "/Users/bbannier/.local/pipx/venvs/zkg/lib/python3.10/site-packages/zeekpkg/manager.py", line 359, in __init__
    make_symlink('packages.zeek', self.autoload_package)
```

```console
$ make -B
rning: skipped using package source named "zeek": failed to clone git repo
make: *** [5.pkg] Error 1
warning: skipped using package source named "zeek": failed to clone git repo
"package-template" will use value of "name" (the name of the package, e.g. "FooBar") from command line: Name
"package-template" will use value of "name" (the name of the package, e.g. "FooBar") from command line: Name
"package-template" will use value of "name" (the name of the package, e.g. "FooBar") from command line: Name
"package-template" will use value of "name" (the name of the package, e.g. "FooBar") from command line: Name
❯ make -B
zkg create --packagedir=1.pkg --user-var name=Name --force
zkg create --packagedir=2.pkg --user-var name=Name --force
zkg create --packagedir=3.pkg --user-var name=Name --force
zkg create --packagedir=4.pkg --user-var name=Name --force
zkg create --packagedir=5.pkg --user-var name=Name --force
zkg create --packagedir=6.pkg --user-var name=Name --force
warning: skipped using package source named "zeek": failed to clone git repo
warning: skipped using package source named "zeek": failed to clone git repo
warning: skipped using package source named "zeek": failed to clone git repo
```

As a workaround for my use case I am using a zkg config file with an empty `sources` section.

--- a/testing/tests/concurrent-state-lock
+++ b/testing/tests/concurrent-state-lock
@@ -1,0 +1,18 @@
+# @TEST-DOC: Concurrent zkg invocations must not race on shared state.
+#
+# @TEST-EXEC: bash %INPUT
+# @TEST-EXEC: btest-bg-wait 30
+#
+# @TEST-EXEC: test ! -s r1/.stderr
+# @TEST-EXEC: test ! -s r2/.stderr
+
+# `btest-bg-run` changes the working directory, so `zkg list` needs an explicit cd to find the test state.
+#
+# This test assumes `zkg list` mutates state, without that the race may not trigger.
+testdir=$(pwd)
+cat > run-zkg-list <<EOF
+cd $testdir && zkg list
+EOF
+
+BTEST_BG_RUN_SLEEP=0 btest-bg-run r1 "sh $testdir/run-zkg-list"
+BTEST_BG_RUN_SLEEP=0 btest-bg-run r2 "sh $testdir/run-zkg-list"

--- a/zeekpkg/manager.py
+++ b/zeekpkg/manager.py
@@ -5,6 +5,7 @@ methods to interact with and operate on Zeek packages.
 
 import configparser
 import copy
+import fcntl
 import filecmp
 import json
 import os
@@ -291,7 +292,13 @@ class Manager:
         self.manifest = os.path.join(self.state_dir, "manifest.json")
         self.autoload_script = os.path.join(self.script_dir, "packages.zeek")
         self.autoload_package = os.path.join(self.script_dir, "__load__.zeek")
+
         make_dir(self.state_dir)
+
+        # The lock file lives inside `state_dir`, so we open it after ensuring the dir exists.
+        self._state_lock = open(os.path.join(self.state_dir, ".lock"), "a")
+        fcntl.flock(self._state_lock, fcntl.LOCK_EX)
+
         make_dir(self.log_dir)
         make_dir(self.scratch_dir)
         make_dir(self.source_clonedir)

